File: c++/best_time_to_buy_and_sell_stock/best_time_to_buy_and_sell_stock.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int lowest_buying_price = prices[0];
        int max_profit = 0;
        for (size_t index = 0; index < prices.size(); index++)
        {
            int price = prices[index];
            if (price < lowest_buying_price)
            {
                lowest_buying_price = price;
            }
            int current_proft = price - lowest_buying_price;
            if (current_proft > max_profit)
            {
                max_profit = current_proft;
            }
        }
        return max_profit;
    }
};

File: java/best_time_to_buy_and_sell_stock/best_time_to_buy_and_sell_stock.java
package best_time_to_buy_and_sell_stock;

class Solution {
    public int maxProfit(int[] prices) {
        int lowest_buying_price = prices[0];
        int max_profit = 0;
        for (int index = 0; index < prices.length; index++) {
            int price = prices[index];
            if (price < lowest_buying_price) {
                lowest_buying_price = price;
            }
            int current_profit = price - lowest_buying_price;
            if (current_profit > max_profit) {
                max_profit = current_profit;
            }
        }
        return max_profit;
    }
}